iron-list.html

```diff
@@ -1362,7 +1410,141 @@
         this._updateMetrics([pidx]);
         this._positionItems();
       }
+    },
+
+    _isIndexRendered: function(idx) {
+      return idx >= this._virtualStart && idx <= this._virtualEnd;
+    },
+
+    _getPhysicalItemForIndex: function(idx, force) {
+      if (!this._collection) {
+        return null;
+      }
+      if (!this._isIndexRendered(idx)) {
+        if (force) {
+          this.scrollToIndex(idx);
+          return this._getPhysicalItemForIndex(idx, false);
+        }
+        return null;
+      }
+      var item = this._getNormalizedItem(idx);
+      var physicalItem = this._physicalItems[this._physicalIndexForKey[this._collection.getKey(item)]];
+
+      return physicalItem || null;
+    },
+
+    _focusPhysicalItem: function(idx) {
+      var physicalItem = this._getPhysicalItemForIndex(idx, true);
+      if (!physicalItem) {
+        return;
+      }
+      var SECRET = ~(Math.random() * 100);
+      var model = physicalItem._templateInstance;
+      var currentTabIndex = model.tabIndex;
+      var focusable;
+
+      model.tabIndex = SECRET;
+      // the focusable element could be the entire physical item
+      if (physicalItem.tabIndex === SECRET) {
+       focusable = physicalItem;
+      }
+      // the focusable element could be somewhere within the physical item
+      if (!focusable) {
+        focusable = Polymer.dom(physicalItem).querySelector('[tabindex="' + SECRET + '"]');
+      }
+      // restore the tab index
+      model.tabIndex = currentTabIndex;
+      focusable && focusable.focus();
+    },
+
+    _manageFocus: function() {
+      var item, pidx;
+      var HIDEEN_Y = '-10000px';
+      var fidx = this._focusedIndex;
+
+      if (this._isIndexRendered(fidx)) {
+        if (this._focusedItem) {
+          item = this._getNormalizedItem(fidx);
+          pidx = this._physicalIndexForKey[this._collection.getKey(item)];
+
+          this.translate3d(0, HIDEEN_Y, 0, this._physicalItems[pidx]);
+          this._physicalItems[pidx] = this._focusedItem;
+          this._focusedItem = null;
+        }
+      } else if (!this._focusedItem) {
+        item = this._getNormalizedItem(fidx);
+        pidx = this._physicalIndexForKey[this._collection.getKey(item)];
+
+        this._focusedItem = this._physicalItems[pidx];
+        this.translate3d(0, HIDEEN_Y, 0, this._focusedItem);
+
+        if (!this._sentinelItem) {
+          var stampedTemplate = this.stamp(null);
+          this._sentinelItem = stampedTemplate.root.querySelector('*');;
+          Polymer.dom(this).appendChild(stampedTemplate.root);
+        }
+        this._physicalItems[pidx] = this._sentinelItem;
+      }
+    },
+
+    _restoreFocusedItem: function() {
+      if (this._focusedItem) {
+        Polymer.dom(this).removeChild(this._focusedItem);

```

Comment:
Can we determine why the `_sentinelItem` needs to be added/removed rather than just moved out of the way when not in use (similar to how `_focusedItem` is moved out of the way when in use)?  It should be cheaper that way than thrashing add/remove.

Also, maybe the naming of `_focusedItem` and `_sentinelItem` could be improved... if I understand correctly (would be good to add this to the comments):
* `_focusedItem` is set to the item that is focused only when and if it is moved offscreen (when it is in the rendered set it is null and only `_focusedIndex` is set)
* `_sentinelItem` is set to the item that backfills the `_focusedItem` in the physical items list when the `_focusedItem` is moved offscreen

So maybe:
* `_focusedItem` :arrow_right: `_offscreenFocusedItem` and
* `_sentinelItem` :arrow_right:  `_focusBackfillItem` ... would be a little clearer? wdyt?